import { plainToInstance } from "class-transformer";
import { validate } from "class-validator";
import { FastifyReply, FastifyRequest } from "fastify";
import { HttpMessages, HttpStatus } from "../config/http.config";
import { LoginDto } from "../dtos/login.dto";
import { RegisterDto } from "../dtos/register.dto";
import { BaseException } from "../exeptions/base.exception";
import { validateDto } from "../middlewares/validate-dto.middleware";
import { AuthService } from "../services/auth.service";
import { ResponseHandler } from "../utils/response.handler";

/**
 * Contr√¥leur pour l'authentification.
 */
export class AuthController {
	private authService: AuthService;

	/**
	 * Cr√©e une instance de AuthController.
	 */
	constructor() {
		this.authService = new AuthService();
	}

	/**
	 * Effectue une connexion.
	 *
	 * @param {FastifyRequest<{ Body: LoginDto }>} request - La requ√™te Fastify.
	 * @param {FastifyReply} reply - La r√©ponse Fastify.
	 * @returns {Promise<FastifyReply>} - La r√©ponse Fastify.
	 */
	async login(
		request: FastifyRequest<{ Body: LoginDto }>,
		reply: FastifyReply,
	): Promise<FastifyReply> {
		try {
			ResponseHandler.info(
				"üîπ Requ√™te re√ßue pour login",
				{ email: request.body.email },
				request,
			);

			// Validation avec `validateDto`
			const loginDto = await validateDto(LoginDto, request.body);

			// Authentification via le service
			const { token, user } = await this.authService.login(
				loginDto,
				request.server,
			);

			ResponseHandler.success("‚úÖ Connexion r√©ussie", user, request);
			request.log.info("üîë Token envoy√© au client (@controller)");

			return reply.status(HttpStatus.OK).send({
				status: "success",
				token,
				user,
			});
		} catch (error: any) {
			if (error instanceof BaseException) {
				ResponseHandler.error("‚ùå Erreur de connexion", error.message, request);
				return reply.status(error.statusCode).send({
					status: "error",
					errorCode: error.statusCode,
					message: error.message,
				});
			}

			ResponseHandler.error("‚ùå Erreur inconnue", error.message, request);
			return reply.status(HttpStatus.INTERNAL_SERVER_ERROR).send({
				status: "error",
				errorCode: HttpStatus.INTERNAL_SERVER_ERROR,
				message: HttpMessages.INTERNAL_SERVER_ERROR,
			});
		}
	}

	/**
	 * Enregistre un nouvel utilisateur.
	 *
	 * @param {FastifyRequest} request - La requ√™te Fastify.
	 * @param {FastifyReply} reply - La r√©ponse Fastify.
	 * @returns {Promise<FastifyReply>} - La r√©ponse Fastify.
	 */
	async register(
		request: FastifyRequest,
		reply: FastifyReply,
	): Promise<FastifyReply> {
		try {
			request.log.info("üì• Requ√™te re√ßue pour inscription", {
				body: request.body,
				requestId: request.id,
			});

			// Validation DTO
			const registerDto = plainToInstance(RegisterDto, request.body);
			const errors = await validate(registerDto);
			if (errors.length > 0) {
				ResponseHandler.error(HttpMessages.BAD_REQUEST, errors, request);
				return reply.status(HttpStatus.BAD_REQUEST).send({
					status: "error",
					errorCode: HttpStatus.BAD_REQUEST,
					message: HttpMessages.BAD_REQUEST,
					errors,
				});
			}

			// Enregistrement de l'utilisateur
			request.log.info("‚úÖ Validation OK, enregistrement en cours...");
			const user = await this.authService.register(registerDto);
			request.log.info("‚úîÔ∏è Inscription r√©ussie", { email: user.email });

			return reply.status(HttpStatus.CREATED).send({
				status: "success",
				message: HttpMessages.CREATED,
				user,
			});
		} catch (error: any) {
			request.log.error("‚ùå Erreur lors de l'inscription", error);
			return reply.status(HttpStatus.BAD_REQUEST).send({
				status: "error",
				errorCode: HttpStatus.BAD_REQUEST,
				message: error.message || HttpMessages.BAD_REQUEST,
			});
		}
	}

	/**
	 * D√©connecte un utilisateur.
	 *
	 * @param {FastifyRequest} request - La requ√™te Fastify.
	 * @param {FastifyReply} reply - La r√©ponse Fastify.
	 * @returns {Promise<FastifyReply>} - La r√©ponse Fastify.
	 */
	async logout(
		request: FastifyRequest,
		reply: FastifyReply,
	): Promise<FastifyReply> {
		try {
			request.log.info("üîì Tentative de d√©connexion");

			const response = await this.authService.logout(request);
			request.log.info("‚úîÔ∏è D√©connexion r√©ussie");

			return reply.status(HttpStatus.OK).send({
				status: "success",
				message: HttpMessages.SUCCESS,
				data: response,
			});
		} catch (error: any) {
			request.log.error("‚ùå Erreur lors de la d√©connexion", error);
			return reply.status(HttpStatus.BAD_REQUEST).send({
				status: "error",
				errorCode: HttpStatus.BAD_REQUEST,
				message: error.message || HttpMessages.BAD_REQUEST,
			});
		}
	}
}
